agent/src/main/java/com/spotify/ffwd/statistics/SemanticCoreStatistics.java

```diff
@@ -246,24 +247,36 @@ public void reportMetricsDroppedByFilter(final int dropped) {
     @Override
     public HighFrequencyDetectorStatistics newHighFrequency() {
         final MetricId m = metric.tagged("component", "high-freq");
-        final AtomicLong highFreqMetrics = new AtomicLong();
+        final Map<MetricId, Long> highFreqMetricsMap = new ConcurrentHashMap<>();
 
         return new HighFrequencyDetectorStatistics() {
             private final Counter dropped = registry.counter(
                 m.tagged("what", "dropped-metrics", "unit", "metric"));
 
-            private final Gauge metricsCardinalityMetric =
-              registry.register(m.tagged("what", "high-freq-metrics"),
-                  (Gauge<Long>) () -> (long) highFreqMetrics.get());
-
             @Override
             public void reportHighFrequencyMetricsDropped(int dropped) {
                 this.dropped.inc(dropped);
             }
 
             @Override
-            public void reportHighFrequencyMetrics(int marked) {
-                highFreqMetrics.set(marked);
+            public void reportHighFrequencyMetrics(int marked, String... tags) {
+                MetricId gaugeMetric = m.tagged("what", "high-freq-metrics").tagged(tags);
+
+                registerGauge(gaugeMetric);
+
+                highFreqMetricsMap.put(gaugeMetric, (long) marked);
+            }
+
+            private synchronized void registerGauge(MetricId gauge) {
+                if (highFreqMetricsMap.containsKey(gauge)) {
```

Comment:
Is this check necessary ? If you want to avoid null value below  return a zero or optional .

Reply:
The registration of new Gauge actually throws an exception if same exactly Gauge is trying to be registered.
As we don't know all possible permutations of different tags of "whats" and "keys" by using synchronized method I'm making sure I register new Gauge only once.